#!/bin/bash
#PBS -P ob80
#PBS -q normal
#PBS -l ncpus=1
#PBS -l mem=4GB
#PBS -l jobfs=6GB
#PBS -l walltime=2:00:00
#PBS -l wd
#PBS -l other=hyperthread
#PBS -l storage=scratch/ob80+gdata/ob80
#PBS -N cutadapt


set -euo pipefail
eval "$(conda shell.bash hook)"
conda activate bioinformatics

mkdir -p cutadapt

if [[ ${R1:-} ]] && [[ ${R2:-} ]]; then
	OUTR1=`basename $R1`
	OUTR2=`basename $R2`
	mkdir -p $PBS_JOBFS/cutadapt
	# can't use paired end mode on fasta files - they need to be fastq format
	cutadapt --cores=0 -b "file:/home/584/re3494/GitHubs/atavide_lite/adapters/IlluminaAdapters.fa" -o $PBS_JOBFS/cutadapt/$OUTR1 $R1
	cutadapt --cores=0 -b "file:/home/584/re3494/GitHubs/atavide_lite/adapters/IlluminaAdapters.fa" -o $PBS_JOBFS/cutadapt/$OUTR2 $R2
	mv -t cutadapt $PBS_JOBFS/cutadapt/$OUTR1 $PBS_JOBFS/cutadapt/$OUTR2
elif [[ ${R1:-} ]]; then
	OUTR1=`basename $R1`
	mkdir -p $PBS_JOBFS/cutadapt
	cutadapt --cores=0 -b "file:/home/584/re3494/GitHubs/atavide_lite/adapters/IlluminaAdapters.fa" -o $PBS_JOBFS/cutadapt/$OUTR1 $R1
	mv -t cutadapt $PBS_JOBFS/cutadapt/$OUTR1
elif [[ ${R2:-} ]]; then
	OUTR2=`basename $R2`
	mkdir -p $PBS_JOBFS/cutadapt
	cutadapt --cores=0 -b "file:/home/584/re3494/GitHubs/atavide_lite/adapters/IlluminaAdapters.fa" -o $PBS_JOBFS/cutadapt/$OUTR2 $R2
	mv -t cutadapt $PBS_JOBFS/cutadapt/$OUTR2
else
	echo "Please define the variables R1 and/or R2. e.g. qsub -v R1=R1.fastq,R2=R2.fasta" >&2; 
fi

