#!/bin/bash
#PBS -P ob80
#PBS -q normal
#PBS -l ncpus=8
#PBS -l mem=32GB
#PBS -l jobfs=20GB
#PBS -l walltime=3:00:00
#PBS -l wd
#PBS -l other=hyperthread
#PBS -l storage=scratch/ob80+gdata/ob80
#PBS -N human


set -euo pipefail
eval "$(conda shell.bash hook)"
conda activate bioinformatics


mkdir -p human no_human
mkdir -p $PBS_JOBFS/human_mapped/ $PBS_JOBFS/human $PBS_JOBFS/no_human


if [[ ${R1:-} ]]; then
	OUTR1=`basename $R1`
	minimap2 -t 8  --split-prefix=$PBS_JOBFS/tmp$$ -a -xsr  ~/databases/human/GCA_000001405.15_GRCh38_no_alt_plus_hs38d1_analysis_set.fna.gz $R1 \
	 | samtools view -bh - > $PBS_JOBFS/human_mapped/$OUTR1.bam

	# reads that match the host
	samtools fasta -F 3588 $PBS_JOBFS/human_mapped/$OUTR1.bam | pigz -c > $PBS_JOBFS/human/$OUTR1

	# reads that don't match the host
	samtools fasta -F 3584 $PBS_JOBFS/human_mapped/$OUTR1.bam | pigz -c > $PBS_JOBFS/no_human/$OUTR1
fi

if [[ ${R2:-} ]]; then
	OUTR2=`basename $R2`
	minimap2 -t 8  --split-prefix=$PBS_JOBFS/tmp$$ -a -xsr  ~/databases/human/GCA_000001405.15_GRCh38_no_alt_plus_hs38d1_analysis_set.fna.gz $R2 \
	 | samtools view -bh - > $PBS_JOBFS/human_mapped/$OUTR2.bam

	# reads that match the host
	samtools fasta -F 3588 $PBS_JOBFS/human_mapped/$OUTR2.bam | pigz -c > $PBS_JOBFS/human/$OUTR2

	# reads that don't match the host
	samtools fasta -F 3584 $PBS_JOBFS/human_mapped/$OUTR2.bam | pigz -c > $PBS_JOBFS/no_human/$OUTR2

fi

rsync -a  $PBS_JOBFS/no_human/  no_human/
rsync -a  $PBS_JOBFS/human/  human/


